lib/SILGen/SILGenApply.cpp

```diff
@@ -1292,7 +1292,9 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
 
     SILValue superMethod;
     auto *funcDecl = cast<AbstractFunctionDecl>(constant.getDecl());
-    if (constant.isForeign || !funcDecl->isFinal()) {
+    auto thisModule = SGF.SGM.M.getSwiftModule();
+    auto inThisModule = thisModule == funcDecl->getModuleContext();
+    if (constant.isForeign || !(funcDecl->isFinal() || inThisModule)) {
```

Comment:
You could factor out these three lines into a new isPolymorphicSuperCall() static function, since the same three lines again appear in the same file.

But the logic should be OK.